### PR TITLE
fix 'y' to true when YAML decode to JSON

### LINKED ISSUE #102308: CRD definition is changing the property name 'y' in openAPIV3Schema spec to True
The below CRD is created using kubectl. The CRD contains the openAPIV3Schema spec which has property name as 'y'.

<pre>
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: tests.calci.io
spec:
  group: calci.io
  versions:
    - name: v1 # it's possible to provide multiple versions of a CRD
      served: true # it's possible to disable a CRD
      storage: true # there can be multiple versions but only one can be used to store the objects
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                operation:
                  type: string
                  enum: ["ADD", "SUB", "DIV", "MUL"]
                x:
                  type: number
                <b>y:
                  type: number</b>
  scope: Namespaced # resources can be namespaced or available for the whole cluster
  names:
    plural: tests
    singular: test
    kind: Test # this name is being used in manifests
    shortNames: # these short names can be used in the CLI, e.g. kubectl get er
    - tt
</pre>

After the crd is created, the y property is showing as True in CRD description

<pre>
Name:         tests.calci.io
Namespace:
Labels:       <none>
Annotations:  API Version:  apiextensions.k8s.io/v1
Kind:         CustomResourceDefinition
Metadata:
  Creation Timestamp:  2021-05-26T05:54:57Z
  Generation:          1
  Managed Fields:
    API Version:  apiextensions.k8s.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:acceptedNames:
          f:kind:
          f:listKind:
          f:plural:
          f:shortNames:
          f:singular:
        f:conditions:
    Manager:      kube-apiserver
    Operation:    Update
    Time:         2021-05-26T05:54:57Z
    API Version:  apiextensions.k8s.io/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        f:conversion:
          .:
          f:strategy:
        f:group:
        f:names:
          f:kind:
          f:listKind:
          f:plural:
          f:shortNames:
          f:singular:
        f:scope:
        f:versions:
      f:status:
        f:storedVersions:
    Manager:         kubectl
    Operation:       Update
    Time:            2021-05-26T05:54:57Z
  Resource Version:  1373282
  UID:               725f0994-7fa1-49b4-a01e-07498b623d2d
Spec:
  Conversion:
    Strategy:  None
  Group:       calci.io
  Names:
    Kind:       Test
    List Kind:  TestList
    Plural:     tests
    Short Names:
      tt
    Singular:  test
  Scope:       Namespaced
  Versions:
    Name:  v1
    Schema:
      openAPIV3Schema:
        Properties:
          Spec:
            Properties:
              Operation:
                Enum:
                  ADD
                  SUB
                  DIV
                  MUL
                Type:  string
              <b>True:
                Type:  number</b>
              X:
                Type:  number
            Type:      object
        Type:          object
    Served:            true
    Storage:           true
Status:
  Accepted Names:
    Kind:       Test
    List Kind:  TestList
    Plural:     tests
    Short Names:
      tt
    Singular:  test
  Conditions:
    Last Transition Time:  2021-05-26T05:54:57Z
    Message:               no conflicts found
    Reason:                NoConflicts
    Status:                True
    Type:                  NamesAccepted
    Last Transition Time:  2021-05-26T05:54:57Z
    Message:               the initial names have been accepted
    Reason:                InitialNamesAccepted
    Status:                True
    Type:                  Established
  Stored Versions:
    v1
Events:  <none>
</pre>

### PR DESCRIPTION
**Pull request by CaoDonghui123**
#### What type of PR is this?

/kind bug
#### Which issue(s) this PR fixes:

Fixes #102308

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go [hunk 1]
```diff
@@ -98,7 +99,7 @@ func (d *YAMLToJSONDecoder) Decode(into interface{}) error {
 	}
 
 	if len(bytes) != 0 {
-		err := yaml.Unmarshal(bytes, into)
+		err := yaml3.Unmarshal(bytes, into)
```
Comment (wzshiming) [inline staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go:102]: Patch this piece of logic separately and introduce new dependencies. I don't think this is a good modification.

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Comment by lavalamp** [comment]
One does not simply increase the version of the yaml library... 😱 

This appears to have broken many unit tests. Ideally we can fix the bug with a much more lightweight tweak.

**Comment by roycaihw** [comment]
```
$ kubectl apply -f crd.yaml -v=9
...
Request Body: {"apiVersion":"apiextensions.k8s.io/v1","kind":"CustomResourceDefinition","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"apiextensions.k8s.io/v1\",\"kind\":\"CustomResourceDefinition\",\"metadata\":{\"annotations\":{},\"name\":\"tests.calci.io\"},\"spec\":{\"group\":\"calci.io\",\"names\":{\"kind\":\"Test\",\"plural\":\"tests\",\"shortNames\":[\"tt\"],\"singular\":\"test\"},\"scope\":\"Namespaced\",\"versions\":[{\"name\":\"v1\",\"schema\":{\"openAPIV3Schema\":{\"properties\":{\"spec\":{\"properties\":{\"operation\":{\"enum\":[\"ADD\",\"SUB\",\"DIV\",\"MUL\"],\"type\":\"string\"},\"true\":{\"type\":\"number\"},\"x\":{\"type\":\"number\"}},\"type\":\"object\"}},\"type\":\"object\"}},\"served\":true,\"storage\":true}]}}\n"},"name":"tests.calci.io"},"spec":{"group":"calci.io","names":{"kind":"Test","plural":"tests","shortNames":["tt"],"singular":"test"},"scope":"Namespaced","versions":[{"name":"v1","schema":{"openAPIV3Schema":{"properties":{"spec":{"properties":{"operation":{"enum":["ADD","SUB","DIV","MUL"],"type":"string"},"true":{"type":"number"},"x":{"type":"number"}},"type":"object"}},"type":"object"}},"served":true,"storage":true}]}}
```
so far it looks like the problem happened in kubectl when sending the request. Could you look into fixing kubectl instead?

**Comment by BenTheElder** [comment]
there's also: https://github.com/kubernetes/kubernetes/issues/78946 which is a bit broader.

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go (hunk 1 above)
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go:102]: `sigs.k8s.io/yaml` uses json tags in structs for marshaling/unmarshaling, and does mechanical yaml-json conversion

`gopkg.in/yaml.v3` expects `yaml` struct tags, which none of our API types have, so this switch will not work at all for API types.

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go [hunk 2]
```diff
@@ -98,7 +99,7 @@ func (d *YAMLToJSONDecoder) Decode(into interface{}) error {
 	}
 
 	if len(bytes) != 0 {
-		err := yaml.Unmarshal(bytes, into)
```
Comment (roycaihw) [inline staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go:101]: this is `sigs.k8s.io/yaml`. Kubernetes depends on the features that `sigs.k8s.io/yaml` implements on top of `go-yaml`. We cannot simply change `sigs.k8s.io/yaml` to `go-yaml` v3.

Please consider moving forward the `go-yaml` v3 conversation in `sigs.k8s.io/yaml`. Some background:
- https://github.com/kubernetes-sigs/yaml/issues/18
- https://github.com/kubernetes-sigs/yaml/pull/26

/close

**Review by roycaihw — COMMENTED** [review]
the fundamental issue is https://github.com/kubernetes/kubernetes/issues/34146. It's unclear if someone will eventually pick it up

**Pull request closed without merging**